contracts/poap/src/contract_tests.rs

```diff
@@ -0,0 +1,962 @@
+        let mut init_msg = get_valid_init_msg(cw721_code_id);
+
+        // Invalid uri
+        init_msg.event_info.base_poap_uri = "invalid_uri".to_string();
+        let init_result = app.instantiate_contract(
+            poap_code_id,
+            Addr::unchecked(ADMIN),
+            &init_msg,
+            &[],
+            "Poap contract",
+            None,
+        );
+        assert!(init_result.is_err());
+
+        // Non ipfs uri
+        init_msg.event_info.base_poap_uri = "https://random_domain.com".to_string();
+        let init_result = app.instantiate_contract(
+            poap_code_id,
+            Addr::unchecked(ADMIN),
+            &init_msg,
+            &[],
+            "Poap contract",
+            None,
+        );
+        assert!(init_result.is_err());
+    }
+
+    #[test]
+    fn instantiate_with_invalid_event_uri() {
+        let mut app = mock_app();
+        let (cw721_code_id, poap_code_id) = store_contracts(&mut app);
+        let mut init_msg = get_valid_init_msg(cw721_code_id);
+
+        // Invalid uri
+        init_msg.event_info.event_uri = "invalid_uri".to_string();
+        let init_result = app.instantiate_contract(
+            poap_code_id,
+            Addr::unchecked(ADMIN),
+            &init_msg,
+            &[],
+            "Poap contract",
+            None,
+        );
+        assert!(init_result.is_err());
+
+        // Non ipfs uri
+        init_msg.event_info.event_uri = "https://random_domain.com".to_string();
+        let init_result = app.instantiate_contract(
+            poap_code_id,
+            Addr::unchecked(ADMIN),
+            &init_msg,
+            &[],
+            "Poap contract",
+            None,
+        );
+        assert!(init_result.is_err());
+    }
+
+    #[test]
+    fn enable_mint() {
+        let (mut app, poap_contract_addr) = proper_instantiate();
+
+        let msg = ExecuteMsg::EnableMint {};
+        app.execute_contract(Addr::unchecked(ADMIN), poap_contract_addr, &msg, &vec![])
+            .unwrap();
+    }
+
+    #[test]
+    fn normal_user_can_t_enable_mint() {
+        let (mut app, poap_contract_addr) = proper_instantiate();
+
+        let msg = ExecuteMsg::EnableMint {};
+        let result = app.execute_contract(Addr::unchecked(USER), poap_contract_addr, &msg, &vec![]);
+        assert!(result.is_err());
+    }
+
+    #[test]
+    fn disable_mint() {
+        let (mut app, poap_contract_addr) = proper_instantiate();
+
+        let msg = ExecuteMsg::DisableMint {};
+        app.execute_contract(Addr::unchecked(ADMIN), poap_contract_addr, &msg, &vec![])
+            .unwrap();
+    }
+
+    #[test]
+    fn normal_user_can_t_disable_mint() {
+        let (mut app, poap_contract_addr) = proper_instantiate();
+
+        let msg = ExecuteMsg::DisableMint {};
+        let result = app.execute_contract(Addr::unchecked(USER), poap_contract_addr, &msg, &vec![]);
+        assert!(result.is_err());
+    }
+
+    #[test]
+    fn mint() {
```

Comment:
It can split into 4 functions
1. mint_not_started_event_poap_error
2. mint_terminated_event_poap_error
3. mint_without_permission_error
4. mint_with_permission_properly (need a state check)